Service/Category/IndexBuilder.php

```diff
@@ -0,0 +1,190 @@
+<?php
+
+namespace Algolia\AlgoliaSearch\Service\Category;
+
+use Algolia\AlgoliaSearch\Exception\CategoryReindexingException;
+use Algolia\AlgoliaSearch\Exceptions\AlgoliaException;
+use Algolia\AlgoliaSearch\Helper\AlgoliaHelper;
+use Algolia\AlgoliaSearch\Helper\ConfigHelper;
+use Algolia\AlgoliaSearch\Helper\Entity\CategoryHelper;
+use Algolia\AlgoliaSearch\Logger\DiagnosticsLogger;
+use Algolia\AlgoliaSearch\Service\AbstractIndexBuilder;
+use Magento\Catalog\Model\ResourceModel\Category\Collection;
+use Magento\Framework\App\Config\ScopeCodeResolver;
+use Magento\Framework\Exception\LocalizedException;
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\Store\Model\App\Emulation;
+
+class IndexBuilder extends AbstractIndexBuilder
+{
+    public function __construct(
+        protected ConfigHelper      $configHelper,
+        protected DiagnosticsLogger $logger,
+        protected Emulation         $emulation,
+        protected ScopeCodeResolver $scopeCodeResolver,
+        protected AlgoliaHelper     $algoliaHelper,
+        protected CategoryHelper    $categoryHelper
+    ){
+        parent::__construct($configHelper, $logger, $emulation, $scopeCodeResolver, $algoliaHelper);
+    }
+
+    /**
+     * @param int $storeId
+     * @param int $page
+     * @param int $pageSize
+     * @return void
+     * @throws LocalizedException
+     * @throws NoSuchEntityException
+     * @throws \Exception
+     */
+    public function rebuildIndex(int $storeId, int $page, int $pageSize): void
+    {
+        if ($this->isIndexingEnabled($storeId) === false) {
+            return;
+        }
+
+        $this->startEmulation($storeId);
+        $collection = $this->categoryHelper->getCategoryCollectionQuery($storeId, null);
+        $this->rebuildIndexPage($storeId, $collection, $page, $pageSize);
+        $this->stopEmulation();
+    }
+
+    /**
+     * @param $storeId
+     * @param $categoryIds
+     * @return void
+     * @throws LocalizedException
+     * @throws NoSuchEntityException
+     * @throws \Exception
+     */
+    public function rebuildIndexIds($storeId, $categoryIds = null): void
```

Comment:
This method name seems a little confusing to me as we are building entities not indices. Some other ideas `buildEntityIds` or just `buildEntities`? 

To really abstract this we could follow the pattern used by Magento indexers and go with:
- `build`
- `buildFull` (calls `build` with empty ids)
- `buildList` (calls `build` with list of entity ids)

These methods could be part of `Api\IndexBuilder`.

Reply:
Since this method is called for updates, I guess we should keep the `rebuild` naming no ? So I'd go for `rebuildEntityIds` if you prefer.